### PR TITLE
Multiple enhancements on proposal's list

### LINKED ISSUE #650: [User details - Proposals] Missing UI feedback when admin censor/approve proposal
- Go to the User details route 
- Click on the proposals tab
- Spam or approve a proposal

Result: Nothing happens (even though the request was successfully completed)

oct-02-2018 14-09-18

This is happening because the proposals displayed on this route aren't updated when the request returns. 
**Note:** The same behavior is happening in the user proposals route.

### PR DESCRIPTION
**Pull request by fernandoabolafio**
closes #650 
- [x] Centralize proposals in one cache in order to have a single source of truth
- [x] Replace approve/censor buttons for loading buttons
- [x] Make sure every route have access to the proposal's vote status

### CODE REVIEW

**Comment by thi4go** [comment]
Tested and everything working great. proposal actions now triggers loading status when needed.

LGTM.

File: src/components/snew/ThingLink.js [hunk 1]
```diff
@@ -99,6 +100,10 @@ const ThingLinkComp = ({
   const loadingStartVote = isRequestingStartVote && startVoteToken === id;
```
Comment (brunobraga95) [inline src/components/snew/ThingLink.js]: :+1:

File: src/reducers/api.js [hunk 2]
```diff
@@ -269,6 +270,53 @@ export const onReceiveVoteStatusChange = (key, newStatus, state, action) => {
   };
 };
 
+export const receiveProposals = (proposals, state) => {
+  const isUnvetted = (prop) =>
```
Comment (brunobraga95) [inline src/reducers/api.js]: :+1:

File: src/reducers/tests/api.test.js [hunk 3]
```diff
@@ -475,6 +497,50 @@ describe("test api reducer", () => {
     ]);
   });
 
+  test("correctly updates the state for onReceiveUserProposals", () => {
```
Comment (brunobraga95) [inline src/reducers/tests/api.test.js]: :100:

File: src/selectors/app.js [hunk 4]
```diff
@@ -181,10 +183,22 @@ export const getDraftProposals = (state) => {
   return drafts;
 };
 
+export const getSubmittedUserProposals = (state) => (userID) => {
+  const isUserProp = prop => prop.userid === userID;
```
Comment (brunobraga95) [inline src/selectors/app.js]: :+1:

File: src/selectors/api.js [hunk 5]
```diff
@@ -255,17 +258,23 @@ export const getPropVoteStatus = state => token => {
 export const userid = state => state.api.me.response && state.api.me.response.userid;
 export const censoredComment = state => state.api.censorComment.response;
 
-const lastLoaded = (func, state) => func(state) ? func(state).lastloaded : {};
+// const lastLoaded = (func, state) => func(state) ? func(state).lastloaded : {};
```
Comment (tiagoalvesdulce) [inline src/selectors/api.js]: I think we can remove these.

File: src/selectors/api.js [hunk 6]
```diff
@@ -255,17 +258,23 @@ export const getPropVoteStatus = state => token => {
...
-  const vp = user(state) ? user(state).proposals : [];
-  const last = Object.keys(vp).length - 1;
-  return last > -1 ? vp[last] : {};
-};
+export const lastLoadedUnvettedProposal = getApiLastLoaded("unvetted");
+export const lastLoadedVettedProposal = getApiLastLoaded("vetted");
+export const lastLoadedUserProposal = getApiLastLoaded("userProposals");
+export const lastLoadedUserDetailProposal = getApiLastLoaded("user");
+
+// export const lastLoadedUserDetailProposal = state => {
```
Comment (tiagoalvesdulce) [inline src/selectors/api.js]: Same as above

**Comment by fernandoabolafio** [comment]
@tiagoalvesdulce this PR is ready for another round of testing. Since this PR changes the proposals flow quite a bit, I recommend you go through the following test cases (beside the set tests you have in your mind): 

- Approve/Censor a proposal on the admin route
- Authorize Vote/Revoke vote/Start Vote on the public proposals route
- Approve/Censor/Authorize Vote/Revoke vote/Start Vote on the user proposals route
- Approve/Censor/Authorize Vote/Revoke vote/Start Vote on the user details route

**Pull request merged**